std: sys: pal: uefi: Overhaul Time

Ayush1325 opened the pull request:
Use UEFI time format for SystemTime.

All calculations and comparisons are being done using UnixTime since UEFI Time format does not seem good fit for those calculations.

I have tested the conversions and calculations, but I am not sure if there is a way to run unit tests for platform specific code in Rust source.

The only real benefit from using UEFI Time representation is that to and fro conversion will preserve `daylight` and `timezone` values.

r? @joboet

Ayush1325 committed:
std: sys: pal: uefi: Overhaul Time

Use a time representation with 1900-01-01-00:00:00 at timezone -1440 min as
anchor. This is the earliest time supported in UEFI.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

fbstj commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -63,9 +142,177 @@ pub(crate) mod system_time_internal {
+        }
+
+        pub(crate) const fn checked_add(&self, dur: Duration) -> Self {
+            let temp: u32 = self.usecs + dur.subsec_nanos();
```
you're combining something called `usecs` with something called `nanos` that doesn't seem to be correct?

Ayush1325 replied:
Isn't usec the name used for nanosecond in unix timestamp: https://docs.rs/libc/latest/libc/struct.timeval.html

Ayush1325 replied:
I have renamed it to `nanos` now. The reason for using `nanos` instead of normal microseconds (which is uses in unix) is because nanoseconds are stored in the UEFI Time normally.

Ayush1325 commented:
@rustbot label +O-UEFI

joboet commented:
This still doesn't solve the problem that `checked_add` and `checked_sub` can return `Some` even though the time does not fit into the bounds of the UEFI time structure. The other problem is that the unspecified timezone is not handled correctly – it's not equal to UTC and comparing/operating on a time with a specified timezone (such as the UNIX epoch) with a time with unspecified timezone is not a defined operation. A panic might be justified in that case.

Then, there is a design question: I think it might be more advantageous to convert the time and date directly into seconds (just like in the old version) as all of the operations use seconds, and it's unlikely that one would create a `SystemTime` without performing arithmetic on it. You'd still need to preserve the timezone information in `SystemTime` though, as it's necessary for bound-checking.

Ayush1325 commented:
> This still doesn't solve the problem that `checked_add` and `checked_sub` can return `Some` even though the time does not fit into the bounds of the UEFI time structure. 

You mean they should fail for bounds outside of those defined here?

```
UINT16 Year; // 1900 - 9999
UINT8 Month; // 1 - 12
UINT8 Day; // 1 - 31
UINT8 Hour; // 0 - 23
UINT8 Minute; // 0 - 59
UINT8 Second; // 0 - 59
UINT32 Nanosecond; // 0 - 999,999,999
INT16 TimeZone; // --1440 to 1440 or 2047
```

> The other problem is that the unspecified timezone is not handled correctly – it's not equal to UTC and comparing/operating on a time with a specified timezone (such as the UNIX epoch) with a time with unspecified timezone is not a defined operation. A panic might be justified in that case.

Ok, but should unspecified to unspecified calculation be fine? Currently the timezone on my qemu also shows up as unspecified, so I think they should be allowed.

> 
> Then, there is a design question: I think it might be more advantageous to convert the time and date directly into seconds (just like in the old version) as all of the operations use seconds, and it's unlikely that one would create a `SystemTime` without performing arithmetic on it. You'd still need to preserve the timezone information in `SystemTime` though, as it's necessary for bound-checking.

Ok, I was thinking the same as all that to and fro conversion for everything was getting a bit too much.

joboet commented:
> > This still doesn't solve the problem that `checked_add` and `checked_sub` can return `Some` even though the time does not fit into the bounds of the UEFI time structure.
> 
> You mean they should fail for bounds outside of those defined here? Or should it fail when the values don't fit in the underlying data structure?

I think we should use the defined bounds (so years from 1900–9999, etc.). `SetTime` is specified to report an error if the time is outside these bounds, and I think it makes sense to respect that.

> 
> > The other problem is that the unspecified timezone is not handled correctly – it's not equal to UTC and comparing/operating on a time with a specified timezone (such as the UNIX epoch) with a time with unspecified timezone is not a defined operation. A panic might be justified in that case.
> 
> Ok, but should unspecified to unspecified calculation be fine? Currently the timezone on my qemu also shows up as unspecified, so I think they should be allowed.

I agree. Only unspecified-specified calculations are undefined.

> > Then, there is a design question: I think it might be more advantageous to convert the time and date directly into seconds (just like in the old version) as all of the operations use seconds, and it's unlikely that one would create a `SystemTime` without performing arithmetic on it. You'd still need to preserve the timezone information in `SystemTime` though, as it's necessary for bound-checking.
> 
> Ok, I was thinking the same as all that to and fro conversion for everything was getting a bit too much.

😄 yeah... I'd maybe still choose a different timebase than UNIX time. You could e.g. use 1900-01-01-00:00[local] as base, which would allow you to use `Duration` again, since all valid times are forward from that date.

Ayush1325 commented:
> 😄 yeah... I'd maybe still choose a different timebase than UNIX time. You could e.g. use 1900-01-01-00:00,0[local] as base, which would allow you to use `Duration` again, since all valid times are forward from that date.

So how should conversion be done? UEFI -> Unix -> Anchor to 1900-01-01-00:00,0 ?
I am not sure how to account for leap years and all that stuff without going through Unix (which has already been documented extensively)

joboet commented:
The source for the algorithm you cite already does a conversion from a timebase of 0000-03-01 to 1970-01-01 by adding/subtracting an offset from the day count.

> The first step in the computation is to shift the epoch from 1970-01-01 to 0000-03-01:
>
> ```
> z += 719468;
> ```


We'd simply need to find the offset for 1900-01-01, use it in the algorithm instead and adjust the UNIX_EPOCH constant accordingly.

joboet commented:
> The source for the algorithm you cite already does a conversion from a timebase of 0000-03-01 to 1970-01-01 by adding/subtracting an offset from the day count.
> 
> > The first step in the computation is to shift the epoch from 1970-01-01 to 0000-03-01:
> > ```
> > z += 719468;
> > ```
> 
> We'd simply need to find the offset for 1900-01-01, use it in the algorithm instead and adjust the UNIX_EPOCH constant accordingly.

By my calculation (well, by inputting the time into the algorithm), the day offset is 693901 for 1900-01-01.

Ayush1325 commented:
> > The source for the algorithm you cite already does a conversion from a timebase of 0000-03-01 to 1970-01-01 by adding/subtracting an offset from the day count.
> > > The first step in the computation is to shift the epoch from 1970-01-01 to 0000-03-01:
> > > ```
> > > z += 719468;
> > > ```
> > 
> > 
> > We'd simply need to find the offset for 1900-01-01, use it in the algorithm instead and adjust the UNIX_EPOCH constant accordingly.
> 
> By my calculation (well, by [inputting the time into the algorithm](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=3a2b86a2aa1591607ff13f4c203dd927)), the day offset is 693901 for 1900-01-01.

I have adjusted both conversion offsets and run some initial tests with time anchored to 1900-01-01. Additionally, I have added checks to ensure that the results of calculations are representable in UEFI Time.

Ayush1325 commented:
ping @joboet

Ayush1325 commented:
> I don't think keeping track of the timezone in `SystemTime` is the right solution here. In particular (AFAICT), the FAT driver in EDK II ignores the timezone (FAT doesn't specify what timezone timestamps are stored in, so Windows and Linux interpret the file times in the current system timezone), meaning that after this PR, setting a file's modification time to two "equal" `SystemTime`s can result in a different actual timestamp being stored if the "equal" `SystemTime`s have different timezones. Instead, I think a better idea would be to always convert the timestamp to UTC before storing it in `SystemTime`, and:

So, the timestamp in `SystemTime::dur` is UTC even right now. Timezone is only used for the following:

1. Converting back to UEFI Time.
2. Comparison between SystemTime. Basically, it is fine to compare times from unspecified timezones, and it is fine to compare between specified timezones, but comparison between unspecified timezone and specified timezone is not allowed.


> 
>     * When fetching the current time, use UTC as the timezone if the timezone is unspecified. This is how EDK II handles it in RTC drivers AFAICT, so would be consistent with the rest of the ecosystem.

Well, that does explain the weird behavior I was observing on OVMF. I can see that I get unspecified timezone, even though I can see the timestamp is UTC.

Why does `RuntimeServices->GetTime` ever return unspecified time to begin with if unspecified is to be treated the same as UTC?

> 
>     * When fetching a file time from the filesystem, use the timezone from the current time if the filesystem file time has an unspecified timezone. This ensures FAT timestamps are interpreted correctly.
> 
>     * When storing a file time to the filesystem, use the timezone from the current time. This ensures FAT timestamps are correctly stored.

Do you have any idea what Project Mu does? If everyone disregards the spec, I guess we might as well. But we probably still need to check.

Ayush1325 commented:
@beetrees Sorry for the late reply. Was busy with college stuff.

I have been pondering things a bit, and I am not sure why we need to do any modification to the time we get from UEFI in Rust. Let's look at Rust's `SystemTime` API:

# Construct a new instance

The only way to do this, is `SystemTime::now`. One can then use calculations on this get a desired time, but there is no way to construct this structure any other way from user API, at least for now. 

Note: If someone calls SetTime, and goes from Unspecified to UTC or vice versa, and then tries to copare to previous time, then I personally would consider it a user error.

# Internal Usage

There might be use cases where we do want to interpret SystemTime Timezone differently, but I feel like that should be on a case by case basis. The base `SysteTime` implementation should not make any assumption regarding this.

Ayush1325 commented:
> A `SystemTime` can also be constructed using `SystemTime::UNIX_EPOCH`. This allows converting between a Unix timestamp and `SystemTime`. For example, the Unix timestamp of a `SystemTime` can be obtained by calling `.duration_since(SystemTime::UNIX_EPOCH)`.

Good point. I have updated the PR.

Ayush1325 committed:
std: sys: pal: uefi: tests: Add systemtime tests

Add tests to ensure that extream system times are still representable.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -50,11 +70,17 @@ impl SystemTime {
+        let temp = Self(self.0.checked_add(*other)?);
+
+        // Check if can be represented in UEFI
+        if temp.to_uefi(0, 0).is_some() { Some(temp) } else { None }
```
Done

Ayush1325 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +92,114 @@ pub(crate) mod system_time_internal {
-    pub(crate) const fn uefi_time_to_duration(t: r_efi::system::Time) -> Duration {
+    pub(crate) const fn from_uefi(t: &Time) -> Duration {
         assert!(t.month <= 12);
         assert!(t.month != 0);
```
Added

Ayush1325 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,125 @@ pub(crate) mod system_time_internal {
+        assert!(t.day <= 31 && t.day != 0);
+
+        assert!(t.second < 60);
+        assert!(t.hour < 24);
```
Added

Ayush1325 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,121 @@ pub(crate) mod system_time_internal {
+        // Check timzone validity
+        assert!(timezone <= 1440 && timezone >= -1440);
+
+        // FIXME: use checked_sub_signed once stablized
```
Added

Ayush1325 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,121 @@ pub(crate) mod system_time_internal {
+            dur.as_secs().checked_add_signed((-timezone as i64) * SECS_IN_MINUTE as i64).unwrap();
+
+        // Convert to UTC
+        let Some(secs) = secs.checked_sub(TIMEZONE_DELTA) else { return None };
```
Thanks for the catch.

phip1611 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,121 @@ pub(crate) mod system_time_internal {
         let month_days: u32 = (m_adj.wrapping_add(adjust) * 62719 + 769) / 2048;
         let leap_days: u32 = y_adj / 4 - y_adj / 100 + y_adj / 400;
-        let days: u32 = y_adj * 365 + leap_days + month_days + (t.day as u32 - 1) - 2472632;
+        let days: u32 = y_adj * 365 + leap_days + month_days + (t.day as u32 - 1) - 2447065;
```
might be worth it to add a `CONST` for the magic `2447065`

phip1611 commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,121 @@ pub(crate) mod system_time_internal {
+        let days = secs / SECS_IN_DAY;
+        let remaining_secs = secs % SECS_IN_DAY;
+
+        let z = days + 693901;
```
more magic constants `693901`, might be worth it to add `CONST` bindings

Ayush1325 commented:
ping @nicholasbishop @joboet

nicholasbishop commented on library/std/src/sys/pal/uefi/tests.rs:
```diff
@@ -23,19 +25,60 @@ fn align() {
-fn epoch() {
-    let t = r_efi::system::Time {
-        year: 1970,
+fn systemtime_start() {
```
Do these tests get run currently? How can I run them locally?

Ayush1325 replied:
Well, I am not sure. I have not yet found any way at least. I run the tests in a seperate rust crate and just copy them here for now.

nicholasbishop replied:
It would be good to add a comment to that effect in this file, otherwise it's not obvious that the tests won't catch errors if you don't manually run them in a separate project.

nicholasbishop commented on library/std/src/sys/pal/uefi/time.rs:
```diff
@@ -66,51 +104,132 @@ pub(crate) mod system_time_internal {
+    /// The changes are to use 1900-01-01-00:00:00 with timezone -1440 as anchor instead of UNIX
+    /// epoch used in the original algorithm.
+    pub(crate) const fn to_uefi(dur: &Duration, timezone: i16, daylight: u8) -> Option<Time> {
+        // Check timzone validity
```
```suggestion
        // Check timezone validity
```

nicholasbishop reviewed (approved):
Lgtm. I'm a little concerned that if the linked blog posts that these conversions are based on go down for some reason, the code will be hard to understand. But I don't think that needs to block the change.

joboet reviewed (approved):
Sorry for the long delay, I've been away from the project for a few weeks. Anyway, this looks good to me, though I'm not sure about the usefulness of `to_uefi` taking a time zone – but we can talk about that when it comes up in practice.

r=me after rebasing

joboet commented:
@bors d+

joboet commented:
@bors delegate+

bors commented:
:v: @Ayush1325, you can now approve this pull request!

If @joboet told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@joboet`
<!-- homu: {"type":"Delegated","delegator":"joboet","delegate":"Ayush1325"} -->

Zalathar commented:
Looking for the cause of https://github.com/rust-lang/rust/pull/145285#issuecomment-3177808046

@bors try jobs=x86_64-gnu-aux
